---
title: "How to talk to your bioinformatician?"
author: 
  - name: "January Weiner"
    orcid: 0000-0003-1438-7819
    affiliations:
     - ref: cubi
affiliations:
  - id: cubi
    name: Core Unit for Bioinformatics, BIH@Charité
    address: Charitéplatz 1, 
    postal-code: 10117 
    city: Berlin 
    country: Germany
title-slide-attributes:
  data-background-image: files/bih_bg_logo.png
format:
  revealjs:
    footer: "Core Unit for Bioinformatics, BIH@Charite"
    theme: cubi.scss
    logo: files/bih_logo_small.png
    transition: fade
    slide-number: "c/t"
    smaller: true
    navigation-mode: linear
    self-contained: true
comments:
  hypothesis: true
knitr:
  opts_chunk:
    dev: "svg"
---

# Who am I to tell you things?

```{scss}
$callout-color-tip: #70ADC1;
```

```{css, echo=FALSE}
.large {
  font-size: 1.5em !important;
  background-color: "#481567" !important;
  color: white !important;
}

.theme-light .mermaid {
    --color-red: #ff7777;
    --color-orange: #ffbb99;
    --color-yellow: #ffeeaa;
    --color-green: #bbffbb;
    --color-cyan: #bbffee;
    --color-blue: #aaccff;
    --color-purple: #ddbbff;
    --color-pink: #ffcccc;
}

.viri1 {
  background-color: "#481567";
  color: white;
}

.whosaid {
  text-align: right !important;
}

.aside {
  font-size: 0.65em !important;
}
```


```{r echo=FALSE,warning=FALSE,message=FALSE}
#| label: setup
library(tidyverse)
library(ggplot2)

theme_set(theme_minimal())
```






## Bioinformatics, statistics, computational biology

```{dot}
#| fig-width: 12
graph G {
  layout = "neato"
  fontname = "Helvetica,Arial,sans-serif"
  scale = 1.2

  node [
		style = filled
		color = "#000000"
		fontname = "Helvetica,Arial,sans-serif"
    fontsize = 10
    shape = "box"
	]

  statistics [ fillcolor="#F8B229", style=filled, label="Statistics", fontsize=18 ]
  bioinformatics [ fontcolor="white", fillcolor="#481567", style=filled, label="Bioinformatics", fontsize=18 ]
  cbio [ fontcolor="white", fillcolor="#33638d", style=filled, label="Computational\nbiology", fontsize=18 ]

  "Computers and science" [ fontcolor="white", fontsize=20, label = "Computers\nand\nScience" ]

  "Computers and science" -- bioinformatics
  "Computers and science" -- cbio
  "Computers and science" -- statistics


  subgraph Bioinformatics {
    node [ style=filled, fillcolor="#481567cc", fontcolor="white", fontsize=14 ]
    bioinformatics -- algorithms -- implementation
    bioinformatics -- formalizations
    bioinformatics -- standards
    bioinformatics -- databases -- maintenance
  }

  subgraph cbio {
      node [ fontcolor="white", fillcolor="#33638dcc", style=filled, fontsize=14 ]
      cbio -- "data\nscience" -- "data\nstewardship"
      cbio -- "functional\ngenomics"
      "functional\ngenomics" -- "gene set\nenrichment"
      cbio -- "using tools"
      cbio -- "sequence\nanalysis" -- "phylogenies"
  }

  subgraph Statistics {
      node [ fillcolor="#F8B22999cc", style=filled, fontsize=14 ]
      statistics -- "hypothesis\ntesting"
      "hypothesis\ntesting" -- "p-values"
      "hypothesis\ntesting" -- "effect sizes"
      statistics -- "power analysis"
      statistics -- "linear\nmodeling" -- "hierarchical\nmodeling"
      statistics -- "visualizations"
      statistics -- "Bayesian\nstatistics"
  }

}
```







::: {.notes}

 * What you call "bioinformatics" are different areas of science
 * We can't know all biology, so be patient with us
 * Bioinformatics is not a service
 * Bioinformatics is a scientific collaboration 

Be patient with us. We are not stupid, but we don't know your context, your
field, your abbreviations.

:::

##

::: {.callout-tip}
## First bottom line

**Talk!**

Keep explaining your project – teach us!

(it goes both ways)

:::

## Things bioinformaticians care about {.incremental}

 * The biological question
 * Statistics
 * Experimental design
 * Quality control
 * Reproducibility
 * Consistency

::: {.notes}
  Differences between "amateur" bioinformaticians like me a while ago and
  "professionals" are mostly in the QC and reproducibility part
:::

# Statistics

## Statistics matters

![](images/statistics.png)

## What is a p-value?

$H_0$: The null hypothesis, no effect

$H_1$: The alternative hypothesis, there is an effect

We run a test, we get a p-value, say $0.03$. It is a probability.

Probability of *what*, exactly?

Raise your hands if you think that the p-value is the...

 . . .


 1. Probability that $H_0$ is true, given the data

 . . .

 1. Probability that $H_1$ is wrong, given the data

 . . .

 1. Probability that the data is random

 . . .

 1. Probability that the observations are due to random chance

 . . .

 1. Probability of getting the same data by random chance

 . . .

 * **Probability of observing an effect at least as extreme given that $H_0$ is true**

::: {.notes}
None of that is true.

P-values are not what we intiutively think they are. 

They are in fact way less useful than we think they are.

The $H_0$ is true or false, there is no probability involved.
:::

## Our intuition is bayesian, not frequentist


| **Frequentist Statistics**    | **Bayesian Statistics**     |
|-------------------------------|-----------------------------|
| ![](images/classicalgiants.png)| ![](images/thomas-bayes.png)      |
| 1. Probability is defined as the long-run frequency of events | 1. Probability represents a degree of belief or certainty about an event |
| 2. Parameters (like the "true value") are fixed but unknown quantities. | 2. Parameters are treated as random variables with their own probability distributions. |
| 3. Asking about the probability of a hypothesis does not make sense | 3. Asking about the probability of a hypothesis is the main goal |

::: {.aside}

Ask me later to show you the coin trick!

:::

## Why is that important?

P-values are the *language* of science, whether we like them (we don't) or
not.

 * *Always* use effect sizes 
 * Never rely on p-values alone
 * Know their limits!

::: {.callout-tip}
## P-values are a language
You have to understand p-values and their limits to talk to other
scientists!
:::

::: {.aside}
Wasserstein RL, Lazar NA. The ASA statement on p-values: context, process, and purpose. The American Statistician. 2016 Apr 2;70(2):129-33.
:::

## How Venn diagrams can fool scientists

COVID-19 study, both COVID-19 patients and non-COVID-19 patients are
compared in two groups of people, *G1* and *G2*. 

We wanted to know whether the influence of COVID-19 is different in these
two groups.

![](images/fig1-1.jpeg)


::: {.aside}
Venn diagrams may indicate erroneous statistical reasoning in
transcriptomics. Weiner, Obermayer and Beule, 2022, Frontiers Genetics
:::


## What happens is, we are comparing significance with non-significance

> The Difference Between “Significant” and “Not Significant” is not
> Itself Statistically Significant

::: {.whosaid}
   (*Andrew Gelman and Howard Stern*)
:::

If a gene is significant in one comparison, and not significant in another,
that does not mean that there is a difference between the two groups.

It simply means that we *failed* to detect the difference in one of the
comparisons, but that is actually quite likely to happen!

::: {.callout-tip}
## Therefore:
Don't say "there is no difference". Say "we did not detect a difference".
:::

::: {.aside}
Gelman A, Stern H. The difference between “significant” and “not significant” is not itself statistically significant. The American Statistician. 2006 Nov 1;60(4):328-31.
:::

## How Venn diagrams can fool scientists

Nieuwenhuis et al. found that half of the scientists who could have
commited this error, did in fact commit this error.

![](images/erroneous_analyses.webp)

::: {.aside}
Nieuwenhuis, Sander, Birte U. Forstmann, and Eric-Jan
Wagenmakers. "Erroneous analyses of interactions in neuroscience: a problem
of significance." Nature neuroscience 14.9 (2011): 1105-1107.
:::


## The results are artifacts!

Groups G1 and G2 were randomly drawn from the same population. They were
not different at all.

![](images/fig2-1.jpeg)

## Going beyond p-values

:::: {.columns}
::: {.column width="45%"}

  * Estimation rather than testing (e.g. confidence intervals rather than
    p-values)
  * Considering effect sizes
  * Power analysis – estimating type II error rates (false negatives)
  * Sign / magnitute errors 
  * Bayesian statistics
  * Correcting for multiple testing


:::
::: {.column width="10%"}
:::
::: {.column width="45%"}

![](images/pregnant.jpg)

:::
::::


# Experimental design

## How many samples are sufficient?

Say, we want to compare two groups with a standard $t-test$, nothing fancy.
Our ability to detect the differences (the statistical *power*) depends on
the sample size and the effect size[^cohen].


```{r echo=FALSE}
#| label: effect_sizes
#| fig-width: 9
#| fig-height: 4
# plot four effect sizes as boxplots
library(tidyverse)
library(ggplot2)
library(ggbeeswarm)

n_samp <- c(3, 4, 5, 7, 10, 15, 25)
d <- c(Large=.8, "Very large"=1.2, Huge=2.0, Ludicrous=4.0)
n_samples <- 100

df <- imap_dfr(d, ~ {
  d_val <- .x
  size <- .y

  x <- rnorm(n_samples, mean = 0, sd = 1)
  y <- rnorm(n_samples, mean = d_val, sd = 1)

  data.frame(value = c(x, y), group = rep(c("A", "B"), each = n_samples), d = d_val, size=size)
}) %>%
  mutate(size = paste0(size, " (d=", d, ")")) %>%
  mutate(size = factor(size, levels=c("Large (d=0.8)", "Very large (d=1.2)", "Huge (d=2)", "Ludicrous (d=4)")))


ggplot(df, aes(x = group, y = value, fill = group)) +
  geom_boxplot(outlier.shape = NA) +
  geom_beeswarm(alpha=.25, corral.width=1.5) +
  facet_wrap(~ size, ncol=4) +
  labs(x = "Group", y = "Value", fill = "Group") +
  theme(legend.position = "none")
```





[^cohen]: Here we use Cohen's $d$

## How many samples are sufficient?

The $y$ axis on this plot shows how the power of the test – meaning how
often, assuming that the groups really differ by $d$ on average, you will
be able to detect the difference using a t-test.

```{r echo=FALSE,cache=TRUE}
#| label: power
# we could use pwr package here, but simulation is more explicit

# parameters
n <- 10000

# simulation
power <- map_dfr(n_samp, ~ {
  n_samples <- .x
  map_dfr(d, ~ {
    d_val <- .x
    pwr <- map_dbl(1:n, ~ {
      x <- rnorm(n_samples, mean = 0, sd = 1)
      y <- rnorm(n_samples, mean = d_val, sd = 1)
      t.test(x, y)$p.value
    }) < 0.05
    data.frame(n_samples = n_samples, d = d_val, power = mean(pwr))
  })
  
}) %>%
  mutate(study = "Simple t-test")
```

```{r}
#| label: power_plot_simple
#| fig-width: 9
#| fig-height: 4
power$n_samples_tot <- 2 * power$n_samples

ggplot(power, aes(x = n_samples, y = power, color = factor(d))) +
  geom_point() +
  geom_line() +
  scale_color_brewer(palette = "Dark2", name="Effect size",
    guide=guide_legend(reverse=TRUE)) +
  scale_x_log10() +
  labs(x = "Number of samples (per group)", y = "Power")

```

**Power**: power of 50% means that on average, you will be able to see
statistical significance in 50% of the experiments *assuming there is a
difference!*

## How many samples are sufficient?

What about the following setup:

 * We have 2 strains (WT and KO)
 * We have treatment + control
 * We want to know whether the treatment has a different effect on the KO
   strain than on the WT strain

This is a 2x2 design, and we need to consider the interaction term.

## How many samples are sufficient?

```{r echo=FALSE,cache=TRUE}
#| label: power_int
# we could use pwr package here, but simulation is more explicit
library(broom)

# parameters
n <- 1000

# simulation
power_int <- map_dfr(n_samp, ~ {
  n_samples <- .x
  map_dfr(d, ~ {
    d_val <- .x
    pwr <- map_dbl(1:n, ~ {
      x <- rnorm(n_samples, mean = 0, sd = 1)
      y <- rnorm(n_samples, mean = 0, sd = 1)
      z <- rnorm(n_samples, mean = 0, sd = 1)
      w <- rnorm(n_samples, mean = 1 * d_val, sd = 1)
      df <- data.frame(x=c(x, y, z, w), 
                       strain=rep(c("WT", "KO"), each=2 * n_samples),
                       group=rep(c("Control", "Treatment"), each=n_samples))
      mod <- lm(x ~ strain * group, data = df)
      anova(mod)$`Pr(>F)`[3]
    }) < 0.05
    data.frame(n_samples = n_samples, d = d_val, power = mean(pwr))
  })
}) %>% 
  mutate(study = "Interaction")
```



```{r echo=FALSE}
#| label: power_plot
power_int$n_samples_tot <- power_int$n_samples * 4

power_tot <- rbind(power, power_int) %>%
  mutate(study = factor(study, levels=c("Simple t-test", "Interaction")))
power_tot$dfac <- factor(power_tot$d)
power_tot$dfac <- factor(power_tot$d, levels=rev(levels(power_tot$dfac)))
ggplot(power_tot, aes(x = n_samples, y = power, color = factor(d))) +
  geom_point() +
  geom_line() +
  scale_color_brewer(palette = "Dark2", name="Effect size",
    guide=guide_legend(reverse=TRUE)) +
  labs(x = "Number of samples (per group)", y = "Power") +
  facet_wrap(~ study, scales = "free_x")
```

## How many samples are sufficient?

That is not even the worse thing.

Simple calculations show that assuming

 * your power is 80% (really great!)
 * $p-value$ cutoff is $0.05$
 * 90% of the $H_0$ are true (i.e., 10% of the time the differences are
   real)

then 36% of your "significant" results **are false positives**[^more]!

(Plus, you failed to detect 20% of the real differences)

[^more]: I can walk you through a very simple demonstration later if you
care.

::: {.aside}
Colquhoun D. An investigation of the false discovery rate and the misinterpretation of p-values. Royal Society open science. 2014 Nov 19;1(3):140216.
:::
 
## 

::: {.callout-tip}
## Bottom line
 Talk to your statistician early!

 Keep your study design simple!
:::

# Is high throughput data worth it?

## {.inverse background-color="#000000"}

:::: {.columns}
::: {.column width="45%"}

![](images/arnolfini.jpg)

:::
::: {.column width="10%"}
:::
::: {.column width="45%"}

::: {.r-stack}
![](images/arnolfini_2.jpg){.fragment}

![](images/arnolfini_3a.png){.fragment}

![](images/arnolfini_3.png){.fragment}

![](images/arnolfini_4.jpg){.fragment}

![](images/arnolfini_5.png){.fragment}
:::

:::
::::

::: {.notes}


:::

## {.inverse background-color="#000000"}

::: {.r-stack}

![](images/monet_4.jpg){.fragment width="800px"}

![](images/monet_3.jpg){.fragment width="800px"}

![](images/monet_2.jpg){.fragment width="900px"}

![](images/monet.jpg){.fragment width="1000px"}


:::

## {.inverse background-color="#000000"}

:::: {.columns}
::: {.column width="45%"}

![](images/monet.jpg)

:::
::: {.column width="10%"}
:::
::: {.column width="45%"}

![](images/sangiorgio.jpg)

:::
::::

## {.inverse background-color="#000000"}

![](images/monet_all.svg)


::: {.notes}

Claude Monet, Saint-Georges-Majeur, 1908, Venice – 37 paintings between 1.10 and 7.12

Almost as many as van Eyck painted during his whole life

:::

## Explorative vs hypothesis testing

:::: {.columns}
::: {.column width="45%"}

#### Explorative analysis

**Pro:**

 * No need to define a-priori hypotheses
 * Something unexpected and new can be found
 * Can be used to generate hypotheses

**Con:**

 * Requires multiple testing correction
 * Requires proper validation
 * Can't do it as the last step

:::
::: {.column width="10%"}
:::
::: {.column width="45%"}

#### Hypothesis-driven analysis

**Pro:**

 * Clear questions
 * Clear answers
 * More statistical power
 * Better story, better paper

**Con:**

 * Requires more planning (and thinking!)
 * Can make you miss something unexpected
 * If you reject the hypothesis, tough luck 

:::
::::

## The bottom line

:::: {.columns}
::: {.column width="45%"}

::: {.callout-tip}
## Do

 * Formulate clear questions

 * Manage your expectations

 * Evaluate existing data - is the approach able to answer your questions?

 * Read papers – which ones are similar to your study?

 * Validate your results

:::

:::
::: {.column width="10%"}
:::
::: {.column width="45%"}

::: {.callout-warning .fragment}
## Don't

 * expect miracles
 
 * "let's just see what we can find"

 * try to save money

 * make too complex designs

:::

:::
::::

# Reproducibility

## Tale of two papers

![](images/pnas_1.png)


## Tale of two papers

![](images/pnas_res_1.png)


## Tale of two papers

![](images/pnas_1.png)

. . .

![](images/pnas_2.png)

## Tale of two papers

![](images/pnas_1.png)


![](images/pnas_2_red.png)

## Tale of two papers

![](images/pnas_res_2.png)

::: {.notes}
Key difference - second paper specifically focused on genes that were
regulated in either mouse model or human data, and asked whether they are
similar or not
:::

## Lessons learned

 * *A lot* depends on how you analyze your data

 . . .

 * This in turn depends on the questions you ask

 . . .

 * The average "Methods" section is not sufficient for reproducible
   science!

> " Second, **none of the 193 experiments** were described in sufficient detail
> in the original paper to enable us to design protocols to repeat the
> experiments, so we had to seek clarifications from the original authors."
> (Errington et al., 2021)

::: {.aside}
Errington TM, Mathur M, Soderberg CK, Denis A, Perfito N, Iorns E, Nosek BA. Investigating the replicability of preclinical cancer biology. Elife. 2021 Dec 10;10:e71601.

Errington TM, Denis A, Perfito N, Iorns E, Nosek BA. Challenges for assessing replicability in preclinical cancer biology. Elife. 2021 Dec 7;10:e67995.
:::


# File formats and data management

## How we work

```{mermaid}
flowchart LR
    A(Excel) --> B(Data import)
    AA(CSV, TSV) --> B(Data import)
    AAA(fastq, ...) --> B(Data import)
    AAAA("Biological\nquestion") --> B(Data import)
    B --> C[Data\ncleanup]
    C --> CC[QC]
    CC --> E[Analysis]
    E --> D
    E --> F(Figures)
    E --> G(Manuscript\nfragments)
    E --> H(Tables\nExcel files)
    F --> I[You]
    G --> I
    H --> I
    I --> J[Fine tuning]
    J --> E
    C --> D[(Long term storage)]
```

## How we work

```{mermaid}
flowchart LR
    A(Excel) --> B(Data import)
    AA(CSV, TSV) --> B(Data import)
    AAA(fastq, ...) --> B(Data import)
    AAAA("Biological\nquestion") --> B(Data import)
    B --> C[Data\ncleanup]
    C --> CC[QC]
    CC --> E[Analysis]
    E --> D
    E --> F(Figures)
    E --> G(Manuscript\nfragments)
    E --> H(Tables\nExcel files)
    F --> I[You]
    G --> I
    H --> I
    I --> J[Fine tuning]
    J --> E
    C --> D[(Long term storage)]
    style C color:#000,fill:#f9f,stroke:#333,stroke-width:4px
    style J color:#000,fill:#f9f,stroke:#333,stroke-width:4px
```

In the diagram above, two things take usually a lot of hands-on time:

 * Understanding and cleaning the data
 * Fine-tuning the analysis results

## Excel and gene names

:::: {.columns}
::: {.column width="60%"}

 * Excel converts some words to dates automatically
 * Gene names like `MARCH1` or `SEPT9` are converted to dates
 * In most cases[^cases], you can't switch off this behavior

[^cases]: You can change the data type for a column to "text" before
pasting data in, but this is just a workaround. In Office 365 it is
possible to switch off this behavior.

:::
::: {.column width="5%"}
:::

::: {.column width="35%"}

:::
::::

## Excel and gene names

::: {.r-stack}
![](images/excel_paper_1.png){.fragment .absolute top=90 left=10 width=80%}

![](images/excel_hgnc.png){.fragment .absolute top=120 left=50 width=80%}

![](images/excel_paper_2.png){.fragment .absolute top=220 left=90 width=80%}
:::


::: {.aside style="font-size: 0.8em; color:red;"}
::: {style="font-size: 0.5em;"}
Ziemann, Mark, Yotam Eren, and Assam El-Osta. "Gene name errors are widespread in the scientific literature." Genome biology 17 (2016): 1-3; 
Abeysooriya M, Soria M, Kasu MS, Ziemann M. Gene name errors: Lessons not learned. PLoS Computational Biology. 2021 Jul 30;17(7):e1008984.
:::
:::

::: {.notes}
2016 paper: 20% of papers have Excel gene name errors
2021 paper: it is 30% now. Excel now learned mulitple languages
:::

## Is Excel suitable for science?

 * How do you record changes?
 * How do you prevent automatic changes?
 * In short – how do you ensure reproducibility?

## How to give us (meta-)data

Part of the communication is passing on the data.

 1. Make sure the data is **complete** (batches? replicates?)
 1. Identifiers should be unique and non-numeric (`ID1` rather than `1`)
 2. Use a separate sheet to describe the meaning of columns
 3. Explain abbreviations
 4. Make the data machine-friendly
 5. Disclose precisely all methods (like models, kit labels etc, request them from service providers!)


## What you should demand from your bioinformaticians

 * Methods used
 * Scripts / pipelines used
 * Full processed data
 * All results as tables (Excel, CSV etc)

Even if you don't know what to do with all that *now*, you might be needing
it in the future!


## How (not to) work with Excel

(for your reference)

 * Avoid manually changing Excel files
 * Never use formatting for data
 * Don't combine values and comments
 * Don't put meta-information into column names
 * One sheet = one table
 * Header = one line
 * Do not use merged cells
 * Use consistent file names
 * Avoid spaces in file and column names (use underscores)


# Some more tips and summaries

## 

:::: {.columns}
::: {.column width="45%"}

### Things we don't like

 * Cleaning up data
 * Data dredging
 * P-hacking
 * Post-hoc hypotheses
 * Excel
 * Manual changes like changing fonts in figures
 * Non-reproducible science

:::
::: {.column width="10%"}
:::
::: {.column width="45%"}

### Things we love

 * Clear questions
 * A priori hypotheses
 * Challenging statistics
 * Creating new tools
 * R and Rmarkdown, or
 * Python and Jupyter
 * Reproducible workflows
 * Well organized data

:::
::::

## Things that you might want to learn

 * Statistics
 * Coding (likely R or Python)
 * Reproducible workflows with Quarto/Rmarkdown or Jupyter

Even if you are not going to use these tools yourself, gaining an insight
into how they work will help you to communicate with your bioinformatician.

## Thank you {.inverse background-color="#000000"}

:::: {.columns}

::: {.column width="40%"}
You can find this presentation along its source code 
at [https://github.com/bihealth/howtotalk](https://github.com/bihealth/howtotalk)

A 5 day R crash course book is available at 
[https://bihealth.github.io/RCrashcourse-book/](https://bihealth.github.io/RCrashcourse-book/)

Course materials & videos:
[https://bihealth.github.io/RCrashcourse2023/](https://bihealth.github.io/RCrashcourse-book/)
 
:::

::: {.column width="10%"}
:::

::: {.column width="50%"}

```{r}
#| label: qr
#| fig-width: 4.5
#| fig-height: 4.5
library(qrcode)
plot(qr_code("https://github.com/bihealth/howtotalk"))
```



:::
::::


## Reproducible workflows with Rmarkdown

```{mermaid}
flowchart LR
    A(Program + Text) -->|knitr| B(Text with\nanalysis results)
    B --> C[LaTeX]
    C --> CC[PDF]
    B --> D[Word]
    B --> E[HTML]
    B --> F[Presentation]
    B --> G[Book]
```

*This can be Rmarkdown, Quarto, Jupyter... the goal is that your code and
your text are in one place, and the results of your calculations are
entered automatically into the text.*

## Reproducible workflows with Rmarkdown


```{r include=FALSE}
p <- 0.05
```

In systems such ar R markdown, you can put directly your
analysis results in your text. For example, when I write that the
$p$-value is equal to `r p`, I am writing this:

```markdown
In systems such ar R markdown, you can put directly your
analysis results in your text. For example, when I write that the
$p$-value is equal to `​r p`, I am writing this:
```

The $p$-value above is not entered manually (as 0.05), but is the result of
a statistical computation. If the data changes, if your analysis changes,
the $p$-value above will automatically change as well.

## Identifiers

 * Never use "pure numerical" identifiers (`1`, `2`, `3`, ...)
 * Never remove columns with "unneeded" identifiers or you can get
   "involuntary anonymization"
 * If you identifier is, say, `S1`, then always refer to that sample as `S1`, not `Smp. 1` or `1`
   or `Sample 1`
 * Better use a unique prefix for a study / cohort / experiment, like
   `RCDB2024_S1`, `RCDB2024_S2`, `RCDB2025_S1`, `RCDB2025_S2`
 * Composite identifiers are fine, as long as you use them consistently:
   `WT_treatment_1`, `KO_control_2` and not `WT_treatment_1`,
   `KO-2-control`, `WT_ctrl2`


::: {.notes}
I have been working in a lab, I know how it is, you can only fit so many
letters on the cap of an Eppendorf tube

 * Identifiers are unique
 * Identifiers are stable
 * Identifiers are machine-readable
:::


## How (not to) work with Excel

:::: {.columns}
::: {.column width="60%"}

### Avoid manually modifying Excel files

 * Manual changes cannot be tracked automatically
 * You have to record every change you make
 * Otherwise, this is not reproducible science!

:::

::: {.column width="5%"}
:::

::: {.column width="35%"}


:::
::::





## How (not to) work with Excel

:::: {.columns}
::: {.column width="60%"}

### Never use formatting for data

Never encode information as formatting, always use explicit columns

Color / font size / font style cannot be read automatically 

:::
::: {.column width="5%"}
:::
::: {.column width="35%"}

![](images/excel_1.png)

:::
::::




## How (not to) work with Excel

:::: {.columns}
::: {.column width="60%"}

### Don't combine values and comments

Make a separate column for comments

Otherwise the values might be lost[^lost]

[^lost]: If the bioinformatician in charge is sloppy

:::
::: {.column width="5%"}
:::
::: {.column width="35%"}

![](images/excel_comments.png)

:::
::::




## How (not to) work with Excel

:::: {.columns}
::: {.column width="60%"}

### Don't put meta-information into column names

Make a separate excel sheet for column meta information

:::
::: {.column width="5%"}
:::
::: {.column width="35%"}

![](images/excel_colnames.png)

:::
::::


## The statistical testing roulette

![](images/salmon_1.png)


## The statistical testing roulette

:::: {.columns}
::: {.column width="45%"}

**Subject.** One mature Atlantic Salmon (Salmo salar) participated in the fMRI study.
The salmon was approximately 18 inches long, weighed 3.8 lbs, and was not alive at
the time of scanning.

**Task.** The task administered to the salmon involved completing an open-ended
mentalizing task. The salmon was shown a series of photographs depicting human
individuals in social situations with a specified emotional valence. The salmon was
asked to determine what emotion the individual in the photo must have been
experiencing.

:::
::: {.column width="10%"}
:::
::: {.column width="45%"}

![](images/salmon_2.png)

:::
::::

::: {.aside}
Bennett CM, Miller MB, Wolford GL. Neural correlates of interspecies perspective taking in the post-mortem Atlantic Salmon: An argument for multiple comparisons correction. Neuroimage. 2009 Jul 1;47(Suppl 1):S125.
:::


